core-rust/mesh-api-server/src/mesh_api/conversions/numerics.rs

```diff
@@ -0,0 +1,67 @@
+use crate::prelude::*;
+
+const MAX_API_STATE_VERSION: u64 = 100000000000000;
+
+pub(crate) fn extract_amount(
+    extraction_context: &ExtractionContext,
+    amount: &models::Amount,
+) -> Result<(ResourceAddress, Decimal), ExtractionError> {
+    let address = ResourceAddress::try_from_bech32(
+        &extraction_context.address_decoder,
+        &amount.currency.symbol,
+    )
+    .ok_or(ExtractionError::InvalidAmount(amount.clone()))?;
+
+    let scale = if amount.currency.decimals < 0 || amount.currency.decimals > 18 {
+        return Err(ExtractionError::InvalidAmount(amount.clone()));
+    } else {
+        dec!(10)
+            .checked_powi(amount.currency.decimals as i64)
+            .unwrap()
+    };
+
+    let quantity = Decimal::from_str(&amount.value)
+        .ok()
+        .and_then(|x| x.checked_div(scale))
+        .ok_or(ExtractionError::InvalidAmount(amount.clone()))?;
+
+    Ok((address, quantity))
+}
+
+pub(crate) fn extract_amount_from_option(
+    extraction_context: &ExtractionContext,
+    amount: Option<Box<crate::mesh_api::generated::models::Amount>>,
+) -> Result<(ResourceAddress, Decimal), ExtractionError> {
+    extract_amount(
+        extraction_context,
+        amount.ok_or(ExtractionError::NotFound)?.borrow(),
+    )
+}
+
+pub fn to_mesh_api_block_index_from_state_version(
+    state_version: StateVersion,
+) -> Result<i64, MappingError> {
+    let state_version_number = state_version.number();
+    if state_version_number > MAX_API_STATE_VERSION {
+        return Err(MappingError::IntegerError {
+            message: "State version larger than max api state version".to_owned(),
+        });
+    }
+    Ok(state_version_number
+        .try_into()
+        .expect("State version too large somehow"))
+}
+
+pub fn to_mesh_api_amount(
```

Comment:
OPTIONAL: Might be nice to have a few mini unit tests here to verify that `extract_amount` and `to_mesh_api_amount` are opposites and can work with e.g. a `currency` at different number of decimals.

Reply:
TODO added